Python/ceval.c

```diff
@@ -3839,19 +3837,20 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
             DEOPT_IF(Py_TYPE(it) != &PyListIter_Type, FOR_ITER);
             STAT_INC(FOR_ITER, hit);
             PyListObject *seq = it->it_seq;
-            if (seq == NULL) {
-                goto iterator_exhausted_no_error;
-            }
-            if (it->it_index < PyList_GET_SIZE(seq)) {
-                PyObject *next = PyList_GET_ITEM(seq, it->it_index++);
-                Py_INCREF(next);
-                PUSH(next);
-                JUMPBY(INLINE_CACHE_ENTRIES_FOR_ITER);
-                NOTRACE_DISPATCH();
+            if (seq) {
+                if (it->it_index < PyList_GET_SIZE(seq)) {
+                    PyObject *next = PyList_GET_ITEM(seq, it->it_index++);
+                    Py_INCREF(next);
+                    PUSH(next);
+                    JUMPBY(INLINE_CACHE_ENTRIES_FOR_ITER);
+                    NOTRACE_DISPATCH();
+                }
+                it->it_seq = NULL;
+                Py_DECREF(seq);
             }
-            it->it_seq = NULL;
-            Py_DECREF(seq);
-            goto iterator_exhausted_no_error;
+            Py_DECREF(POP());
+            JUMPBY(INLINE_CACHE_ENTRIES_FOR_ITER + oparg);
+            DISPATCH();
```

Comment:
I believe this can be `NOTRACE_DISPATCH()` now, right? Likewise in FOR_ITER_RANGE

Reply:
I honestly have a hunch that a bunch of the existing `NOTRACE_DISPATCH()` usage is incorrect and could trigger assertion failures (but haven't had time to look into it yet). I don't think `NOTRACE_DISPATCH()` is safe if we decref an untrusted value (like a list containing arbitrary elements) since the destructor could start tracing. Right?

So in the `range` case it's safe, but probably not the `list` case (if my hunch is correct). What do you think?

Reply:
I'm curious how much perf win we're getting from the `NOTRACE_DISPATCH()` stuff. It seems really tricky to get right, and it only saves one arithmetic operation on `oparg`.

Reply:
I thought we were assuming that `__del__` methods execute at an unspecified time, so if they turn on tracing, it shouldn't matter whether we actually start tracing a few instructions later.

I guess the unfortunate scenario would be where a `__del__` method turns on tracing, then sees a few instructions traced, then returns to the Py_DECREF call site, and then can take a couple of instructions before it starts tracing that caller. I'm not sure whether that is acceptable.

Reply:
Hm, well in that case we should at least remove all those assertions at the start of the quickened instructions. I also want to try benchmarking a branch were we just use `DISPATCH()` for everything, and see how much slower that is.

Reply:
#96636